Feat/investment potential logic

Voronapxl opened the pull request.

Voronapxl committed:
feat(investment_logic): added logic for investment calculations by territory, and commented for functional zones

Voronapxl committed:
feat(functional_zones_ia): featured logic for calculating investment attractivnes with functional zones

Voronapxl committed:
feat(investment_logic): added method for investment calculations when custom coordinates are presented and adjusted logic for indicators assigment

LeonDeTur commented on .gitignore:
```diff
@@ -125,7 +125,7 @@ celerybeat.pid
 
 # Environments
-app/.env.development
+.env.development
```
.env.*

Voronapxl replied:
done

LeonDeTur commented on app/urbanomy_api/dto/investment_attractivness_dto.py:
```diff
@@ -0,0 +1,115 @@
+}
+
+
+class InvestmentAttractivenessQueryDto(BaseModel):
+    scenario_id: int = Field(..., example=198, description="Scenario ID")
+    to_return: str = Field(..., example="geodataframe", description="Which format to return")
+
+
+class InvestmentAttractivenessBodyDto(BaseModel):
+    benchmarks: Dict[str, Dict[str, Any]] = Field(
+        ...,
+        example=benchmarks_demo,
+        description="Dictionary of benchmarks data for each land use category"
+    )
```
переписать под одно дто для каждого запроса и описать модели для bencmark_demo (можно даже с deafault)

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,225 @@
+
+class InvestmentPotentialService:
+    @staticmethod
+    async def calculate_landuse_score(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame | pd.DataFrame:
```
только gdf

Voronapxl replied:
done

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,225 @@
+            return gdf.to_crs(gdf.estimate_utm_crs())
+
+        records: list[dict] = []
+        for _, row in gdf.iterrows():
+            for ip_type, col in LAND_USE_TO_POTENTIAL_COLUMN.items():
+                records.append({
+                    "ip_type": ip_type,
+                    "ip_value": row[col],
+                    "geometry": row.geometry
+                })
```
переписать под apply

Voronapxl replied:
done

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,225 @@
+            gdf[name] = value
+        records = []
+        for _, row in gdf.iterrows():
+            for ip_type, col_name in LAND_USE_TO_POTENTIAL_COLUMN.items():
+                records.append({
+                    "ip_type": ip_type,
+                    "ip_value": row[col_name],
+                    "geometry": row.geometry
+                })
```
тоже apply

Voronapxl replied:
done

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,225 @@
+        mapped_zones_gdf = mapped_zones_gdf.to_crs(mapped_zones_gdf.estimate_utm_crs())
+        gdf_out, summary = await InvestmentPotentialService.calculate_investment_attractiveness(mapped_zones_gdf,
+                                                                                                benchmarks)
+        if to_return == "dataframe":
+            response = await InvestmentPotentialService.generate_response(summary, to_return)
+            return response
+        elif to_return == "geodataframe":
+            response = await InvestmentPotentialService.generate_response(gdf_out, to_return)
+            return response
```
лишнее, в generate response есть

Voronapxl replied:
done

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,225 @@
+        mapped_zones_gdf = mapped_zones_gdf.to_crs(mapped_zones_gdf.estimate_utm_crs())
+        gdf_out, summary = await InvestmentPotentialService.calculate_investment_attractiveness(mapped_zones_gdf,
+                                                                                                benchmarks)
+        if to_return == "dataframe":
+            response = await InvestmentPotentialService.generate_response(summary, to_return)
+            return response
+        elif to_return == "geodataframe":
+            response = await InvestmentPotentialService.generate_response(gdf_out, to_return)
+            return response
```
лишнее, в generate response есть

Voronapxl replied:
done

LeonDeTur commented on app/urbanomy_api/modules/urban_api_gateway.py:
```diff
@@ -73,14 +75,60 @@ async def get_functional_zones(scenario_id: int, is_context: bool = False, sourc
+        geometries = []
+        for feature in features:
+            try:
+                geom = shape(feature["geometry"])
+                if not geom.is_valid:
+                    geom = geom.buffer(0)
+                geometries.append(geom)
+            except Exception as e:
+                logger.error(f"Error processing geometry: {e}")
+                geometries.append(None)
```
просто gpd.GeoDataFrame.from_features

Voronapxl replied:
gotovo

LeonDeTur commented on app/urbanomy_api/urbanomic_controller.py:
```diff
@@ -0,0 +1,56 @@
+    except ValueError as e:
+        raise http_exception(400, str(e))
+
+    return result
```
унести обработку ошибок в сервис

LeonDeTur commented on requirements.txt:
```diff
@@ -1,9 +1,17 @@
+geojson_pydantic~=2.0.0
+numpy
+pandera
+matplotlib
```
указать версии
пандера прекратила поддержу pandas, так что вообще сломаться может, если ты для pandas её используешь

Voronapxl replied:
При старте приложения оно валится с такой вот ошибкой без панедры при импорте 

>   File "F:\Projects\UrbanomyAPI\.venv\Lib\site-packages\urbanomy\methods\investment_potential\__init__.py", line 8, in <module>
>     from .land_use_score import LandUseScoreAnalyzer
>   File "F:\Projects\UrbanomyAPI\.venv\Lib\site-packages\urbanomy\methods\investment_potential\land_use_score.py", line 8, in <module>
>     from pandera import check_types
> ModuleNotFoundError: No module named 'pandera'

Voronapxl replied:
на самом деле не уеврен, это опять у меня где-то в нейминге ошибка или со стороны либы урабномии

LeonDeTur replied:
укажи версию из либы или убери тогда из реков, либа сама подтянет

LeonDeTur reviewed (changes requested):
Исправить
Главное дто и сервис
и json | geojson логику поправить

Voronapxl committed:
fix(investment_logic): additional logic fixes

Voronapxl committed:
fix(DTO_fixes): added separate models for benchmark dicts, changed return format to bool and renamed to "as_geojson"

Voronapxl committed:
fix(geom_DTO): deprecated usage of gejson pydantic and added a custom model

LeonDeTur commented on app/urbanomy_api/dto/investment_attractivness_dto.py:
```diff
@@ -2,48 +2,51 @@
 from pydantic import BaseModel, Field
 
-benchmarks_demo = {
+residential_demo: Dict[str, Dict[str, Any]] = {
```
Унеси это мапу всю в json, толку от неё в файле, только мешается

LeonDeTur commented on app/urbanomy_api/dto/investment_attractivness_dto.py:
```diff
@@ -92,24 +95,44 @@
+class InvestmentAttractivenessRequestDto(BaseModel):
+    scenario_id: int = Field(..., example=198, description="ID сценария")
+    as_geojson: bool = Field(..., example=False, description="Which format to return")
+    residential: Dict[str, ResidentialBenchmarkDto] = Field(
```
ну вообще в идеале как будто и под это описать модель
ну или может быть хотя бы через Dict[Literal["key_1", ["key_2"], ..., ResidentialBenchmarkDemo]]
Так непонятно, обязательные ли поля и можно передать свои какие-то или нет

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -163,9 +163,9 @@ def _resolve_value(zid: int, itype: str) -> float | None:
             gdf_out: gpd.GeoDataFrame,
             summary: pd.DataFrame,
-            to_return: str
+            to_return: bool = False
```
переименуй флаг, непонятно что значит
as_geojson поставь или return_as_geojson

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -194,11 +194,11 @@ async def run_investment_calculation(scenario_id, to_return: str, benchmarks: di
-    async def run_investment_calculation_fzones(scenario_id, to_return: str, benchmarks: dict[str, dict[str, any]]) \
+    async def run_investment_calculation_fzones(scenario_id, as_geojson: bool, benchmarks: dict[str, dict[str, any]]) \
             -> gpd.GeoDataFrame | pd.DataFrame:
         logger.info(f"Running investment calculation for scenario {scenario_id}")
```
benchmarks тоже в логи добавь
а лучше выводи логи с параметрами запроса выше

LeonDeTur commented on app/urbanomy_api/urbanomic_controller.py:
```diff
@@ -14,27 +14,34 @@
-    except ValueError as e:
-        raise http_exception(400, str(e))
-
+    benchmarks_dict = {
+        **{k: v.dict() for k, v in params.residential.items()},
+        **{k: v.dict() for k, v in params.non_residential.items()}
+    }
```
унести в модель под функцию c декоратором @field_validator
https://docs.pydantic.dev/latest/concepts/validators/#using-the-annotated-pattern

LeonDeTur commented on app/urbanomy_api/urbanomic_controller.py:
```diff
@@ -14,27 +14,34 @@
-        body: InvestmentAttractivenessBodyDto = Body(...)
+        params: Annotated[InvestmentAttractivenessRequestDto, Body(...)],
 ):
+    benchmarks_dict = {
+        **{k: v.dict() for k, v in params.residential.items()},
+        **{k: v.dict() for k, v in params.non_residential.items()}
+    }
```
тоже в модель

LeonDeTur commented on app/urbanomy_api/urbanomic_controller.py:
```diff
@@ -44,13 +51,13 @@ async def calculate_investment_attractiveness_functional_zones(
 async def calculate_investment_attractiveness_by_coords(
-        dto: InvestmentAttractivenessCoordsDto = Depends(),
-        body: InvestmentAttractivenessCoordsBody = Body(...),
+        dto: Annotated[InvestmentAttractivenessCoordsDto, Body(...)],
```
Почему Body, а не Depends?

LeonDeTur commented on app/urbanomy_api/schemas/features_model.py:
```diff
@@ -0,0 +1,134 @@
+        return {"type": self.type, "coordinates": self.coordinates}
+
+
+class Feature(BaseModel):
```
добавить field_validator для обязательных пропертей

LeonDeTur commented on requirements.txt:
```diff
@@ -1,9 +1,17 @@
+shapely~=2.1.1
+numpy
+pandera
+matplotlib
```
версии указать для остальных

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,233 @@
+    @staticmethod
+    async def run_investment_calculation(scenario_id, as_geojson: bool, benchmarks: dict[str, dict[str, any]]) \
+            -> gpd.GeoDataFrame | pd.DataFrame:
+        logger.info(f"Running investment calculation for scenario {scenario_id}")
```
логер прям на запрос накинуть в сервисе
обернуть в try-except как минимум либу, а как максимум места, где потенциально может упасть

LeonDeTur reviewed (changes requested):
ДТО доделать
Все методы, которые могут упасть обернуть в try-except
добавить версионность либ (gunicorn точно)

Voronapxl committed:
feat(DTO_features):
- Extract benchmark DTOs into a separate module
- Move benchmark example data into its own JSON file
- Update parameter-passing logic in all endpoints
- Fix Feature model example and add a validator for mandatory properties
- Pin dependency versions in requirements.txt

LeonDeTur commented on app/urbanomy_api/dto/benchmarks_dto.py:
```diff
@@ -0,0 +1,51 @@
+                **residential_demo,
+                **non_residential_demo
+            }
+        }
```
переименуй, вдруг тебе потом понадобится класс Config из IDU-config

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,216 @@
+            for rec in records:
+                new_rec = {}
+                for k, v in rec.items():
+                    if isinstance(v, float) and math.isnan(v):
```
проверяй на pd.isna() у него больше покрытие вариантов nan

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,216 @@
+            score_gdf = analyzer.compute_scores_long(gdf)
+            logger.info(f"landuse score have been calculated")
+        except Exception:
+            raise http_exception(500, "Error calculating landuse score")
```
input и detail добавить
и logger.exception(e)

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,216 @@
+                axis=1
+            )
+        except Exception:
+            raise http_exception(500, "Error adjusting indicator values")
```
логер, инпут, детэил

LeonDeTur commented on app/urbanomy_api/modules/invest_potential_service.py:
```diff
@@ -0,0 +1,216 @@
+
+        long_df = pd.DataFrame(records)
+        long_gdf = gpd.GeoDataFrame(long_df, geometry="geometry", crs=gdf.crs)
+        logger.info("Indicator values fetched successfully")
```
указать для чего фетч удался

LeonDeTur commented on app/urbanomy_api/urbanomic_controller.py:
```diff
@@ -0,0 +1,50 @@
+        params: Annotated[InvestmentAttractivenessCoordsDto, Depends(InvestmentAttractivenessCoordsDto)],
+):
+    benchmarks_dict: Dict[str, Dict[str, Any]] = params.benchmarks.model_dump()
+    result = await InvestmentPotentialService.run_investment_calculation_coords(
```
прокидывай прям парамс, меньше строчек и переменных

Voronapxl replied:
Пока предлагаю оставить так, там путаннца с тем, что он будет пытаться передать экземпляр класса

LeonDeTur replied:
кк, потом

LeonDeTur reviewed (changes requested):
Вот это доисправь и апрувлю
в целом отлично

Voronapxl committed:
fix(investment_potential_service): comment fixes

Voronapxl committed:
fix(benchmarks_dto): renamed Config

LeonDeTur reviewed (approved):
версию пандеры поправить или убрать из зависимостей

Voronapxl committed:
fix(dependencies): fixed requirements and python version

Voronapxl committed:
fix(dto): added correct argument name
